src/main/java/com/astrapay/qris/cpm/QrCpmParser.java

```diff
@@ -1,5 +1,153 @@
 package com.astrapay.qris.cpm;
 
+import com.astrapay.qris.QrisHexConverter;
+import com.astrapay.qris.cpm.enums.DataType;
+import com.astrapay.qris.cpm.enums.TagIndicator;
+import com.astrapay.qris.cpm.object.QrCpmDataObject;
+import com.astrapay.qris.cpm.object.QrCpmPayload;
+import lombok.Setter;
+import org.apache.commons.codec.DecoderException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+@Service
 public class QrCpmParser {
     // function to parse qr cpm
+
+    @Autowired
+    private QrisHexConverter qrisHexConverter;
+
+    private static final Integer HEX_RADIX = 16;
+
+    private static final Integer SIZE_TO_EXTRACT_LENGTH = 2;
+    private static final Integer HEX_PAIR_REPRESENTATION = 2;
+    private static final Integer START_INDEX_TO_CONVERT_COMPRESSED_NUMERIC = 0;
+
+    @Autowired
+    @Setter
+    private List<String> qrisCpmSubTag;
+
+    @Autowired
+    @Setter
+    private List<String> applicationTemplateSubTag;
+
+    @Autowired
+    @Setter
+    private List<String> applicationSpecificTransparentTemplateSubTag;
+
+    public QrCpmPayload parse(String payloadBase64) throws IOException, DecoderException {
+
+        byte[] payloadByteArrayBased = qrisHexConverter.decodeFromBase64(payloadBase64);
+        String payloadHexBased = qrisHexConverter.encode(payloadByteArrayBased);
+
+        QrCpmPayload qrCpmPayload = new QrCpmPayload();
+        qrCpmPayload.setPayloadBase64(payloadBase64);
+        qrCpmPayload.setPayloadHex(payloadHexBased);
+
+        parse(qrCpmPayload);
+
+        return qrCpmPayload;
+    }
+
+    private void parse(QrCpmPayload qrCpmPayload) throws DecoderException {
+        Map<String, QrCpmDataObject> qrisCpmMap = new LinkedHashMap<>();
+        parseRoot(qrCpmPayload.getPayloadHex(), qrisCpmMap);
+        parseApplicationTemplate(qrisCpmMap);
+        parseApplicationSpecificTransparentTemplate(qrisCpmMap);
+
+        qrCpmPayload.setQrisRoot(qrisCpmMap);
+    }
+
+    private void parser(List<String> qrisCpmSubTag, String payloadHexBased, Map<String, QrCpmDataObject> qrisCpmMap) throws DecoderException {
```

Comment:
Please add comments about what are this method do and why.
Can be added here or inside its scope (ahead of the while-scope and ahead of the if-scope).